client/lib/purchases/index.ts

```diff
@@ -41,51 +41,165 @@ import type { ReactChild } from 'react';
 
 const debug = debugFactory( 'calypso:purchases' );
 
-export interface SiteWithPurchases {
+interface PurchaseWithStatus extends Purchase {
+	currentPurchaseStatus: string;
+}
+
+interface PurchaseStatus {
+	[ key: string ]: number;
+}
+
+interface SiteStatus {
+	[ key: string ]: number;
+}
+
+interface SiteWithPurchases {
 	id: number;
 	name: string;
 	slug: string;
 	isDomainOnly: boolean;
 	title: string;
-	purchases: Purchase[];
+	purchases: PurchaseWithStatus[];
+	isConnected: boolean;
 	domain: string;
+	currentStatus?: string;
 }
+
 export type TracksProps = Record< string, string | number | boolean >;
 
+// Site Sort order: (all purchases are grouped by site) - Sorted by if
+// the site or purchases of the site require some action.
+const siteStatus: SiteStatus = {
+	disconnected: 10,
+	pendingActivation: 20,
+	hasExpired: 30,
+	hasExpiringSoon: 40,
+	hasPaymentMethodExpired: 50,
+	noPaymentActionNeeded: 60,
+	hasCannotManage: 100,
+};
+
+// Sort order of the Purchases of each site (sorted by purchase status)
+const purchaseStatus: PurchaseStatus = {
+	expired: 10,
+	paymentMethodExpired: 20,
+	expiringSoon: 30,
+	noPaymentActionNeeded: 40,
+	cannotManage: 100,
+};
+
+function getSitePurchasesStatus( site: SiteWithPurchases ) {
+	if ( ! site.isConnected ) {
+		if ( site.slug === 'siteless.jetpack.com' ) {
+			return 'pendingActivation';
+		}
+		return 'disconnected';
+	}
+	const { purchases } = site;
+
+	if ( purchases.some( ( purchase ) => purchase.currentPurchaseStatus === 'cannotManage' ) ) {
+		return 'hasCannotManage';
+	}
+	if ( purchases.some( ( purchase ) => purchase.currentPurchaseStatus === 'expired' ) ) {
+		return 'hasExpired';
+	}
+	if (
+		purchases.some( ( purchase ) => purchase.currentPurchaseStatus === 'paymentMethodExpired' )
+	) {
+		return 'hasPaymentMethodExpired';
+	}
+	if ( purchases.some( ( purchase ) => purchase.currentPurchaseStatus === 'expiringSoon' ) ) {
+		return 'hasExpiringSoon';
+	}
+
+	return 'noPaymentActionNeeded';
+}
+
+function getPurchaseStatus( purchase: PurchaseWithStatus ) {
+	const expiry = moment( purchase.expiryDate );
+
+	if ( purchase.isInAppPurchase || isPartnerPurchase( purchase ) ) {
+		return 'cannotManage';
+	}
+	if ( isExpired( purchase ) ) {
+		return 'expired';
+	}
+	if ( creditCardHasAlreadyExpired( purchase ) ) {
+		return 'paymentMethodExpired';
+	}
+	if (
+		( isExpiring( purchase ) &&
+			expiry < moment().add( 30, 'days' ) &&
+			! isRecentMonthlyPurchase( purchase ) ) ||
+		( isRenewing( purchase ) &&
```

Comment:
Should that second part of the OR return `'paymentMethodExpired'` instead?

Reply:
No, I don't believe so.
The first part of the OR is checking if the purchase is set to "manual renew" and it's expiring soon.  The second part of the OR is checking if the purchase is an "auto renew" and it's expiring soon, because when a purchase is set to 'auto renew', it won't ever expire, unless the payment method ends up expiring before the renew date.  That's what the second part of this OR is checking: it's saying, if the purchase is 'auto-renewing' and it has a renew date, and the creditCard expires before the renew date, then it's 'expiring soon'. 😉